File: 10regularExpressionMatching/regularExpressionMatching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    s = [...s]
    p = [...p]
    let dp = [true, ...Array(p.length).fill(false)]
    for (let column = 2; column < dp.length; ++column) dp[column] = dp[column - 2] && Object.is(p[column - 1], '*')
    for (let row = 1; !Object.is(row, s.length + 1); ++row)
    {
        let prev = [...dp]
        dp[0] = false
        for (let column = 1; !Object.is(column, dp.length); ++column)
        {
            if (Object.is(s[row - 1], p[column - 1]) || Object.is(p[column - 1], '.')) dp[column] = prev[column - 1]
            else if (Object.is(p[column - 1], '*')) dp[column] = dp[column - 2] || prev[column] && (Object.is(s[row - 1], p[column - 2]) || Object.is(p[column - 2], '.'))
            else dp[column] = false
        }
    }
    return dp[dp.length - 1]
};

File: 10regularExpressionMatching/regularExpressionMatching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        std::vector<bool> dp(std::size(p) + 1);
        dp.front() = true;
        for (std::size_t column{2}; column < std::size(dp); ++column) dp.at(column) = dp.at(column - 2) && p.at(column - 1) == '*';
        for (std::size_t row{1}; row != std::size(s) + 1; ++row)
        {
            auto const prev{dp};
            dp.front() = false;
            for (std::size_t column{1}; column != std::size(dp); ++column)
            {
                if (s.at(row - 1) == p.at(column - 1) || p.at(column - 1) == '.') dp.at(column) = prev.at(column - 1);
                else if (p.at(column - 1) == '*') dp.at(column) = dp.at(column - 2) || prev.at(column) && (p.at(column - 2) == s.at(row - 1) || p.at(column - 2) == '.');
                else dp.at(column) = false;
            }
        }
        return dp.back();
    }
};
